### PR TITLE
[SPARK-21045][PYSPARK]Fixed executor blocked because traceback.format_exc throw UnicodeDecodeError

### PR DESCRIPTION
**Pull request by dataknocker**
## What changes were proposed in this pull request?

when encode throw exception, use origin exception info.

## How was this patch tested?

We can run in pyspark:
spark = SparkSession.builder.master('local').getOrCreate()
rdd = spark.sparkContext.parallelize(['中']).map(lambda x: x.encode("utf8"))
rdd.count()

Before fixed this bug, this program will be blocked.
After fixed this bug, this program will throw exception expected.

### CODE REVIEW

**Comment by dataknocker** [comment]
We can choose one pr in this and [#18261](https://github.com/apache/spark/pull/18261) to slove SPARK-21045

File: python/pyspark/worker.py [hunk 1]
```diff
@@ -177,8 +177,13 @@ def process():
             process()
     except Exception:
         try:
+            exc_info = traceback.format_exc()
+            try:
+                exc_info = exc_info.encode("utf-8")
```
Comment (HyukjinKwon) [inline python/pyspark/worker.py]: I think we should not just rely on the exception handling but only need to `encode` on `unicode` in Python 2 and `str` in Python 3.

**Comment by HyukjinKwon** [comment]
Also, I think we need a small test to check if it really fixes the problem.

File: python/pyspark/worker.py (hunk 1 above)
Comment (dataknocker) [inline python/pyspark/worker.py]: In Python 2, traceback.format_exc() is already str, don't need to encode. So can I change like this:
<pre>
<code>
    exc_info = traceback.format_exc()
    if sys.version > '3':
        exc_info = exc_info.encode("utf-8")
    write_int(SpecialLengths.PYTHON_EXCEPTION_THROWN, outfile)
    write_with_length(exc_info, outfile)
</code>
</pre>

Comment (HyukjinKwon) [inline python/pyspark/worker.py]: I would rather go for

```python
if sys.version >= '3':
    unicode = str
...
if isinstance(exc_info, unicode):
    exc_info = exc_info.encode('utf-8'))
else:
    exc_info = exc_info.str(desc)
  
```

approach for sure (at least for sanity check).

Comment (dataknocker) [inline python/pyspark/worker.py]: Benefit a lot, thx.

exc_info = exc_info.str(desc)    is right? I think I can remove it.

```python
if sys.version >= '3':
    unicode = str
if isinstance(exc_info, unicode):
    exc_info = exc_info.encode('utf-8'))

```

Comment (HyukjinKwon) [inline python/pyspark/worker.py]: Yea, we could refer `https://github.com/apache/spark/pull/17267`.

**Comment by HyukjinKwon** [comment]
@dataknocker, probably, you are able to add a test something like this:

```python
def e1():
    raise Exception(u"아".encode("UTF-8"))

spark.sparkContext.parallelize([1]).map(lambda x: e1()).collect()
```

**Comment by dataknocker** [comment]
@HyukjinKwon ok, I will add my test later.

**Comment by HyukjinKwon** [comment]
cc @viirya here too. (This one is similar one with few PRs you and I are reviewing).

**Comment by dataknocker** [comment]
Some thing wrong, so I close this pr and add new pr [#18324](https://github.com/apache/spark/pull/18324)

**Pull request closed without merging**